syntax = "proto3";
package LabNet.Server;
import "LabNet.proto";
import "google/protobuf/timestamp.proto";


// general messages

message DigitalOutState {
	PinId pin = 1;
	bool state = 2;
	google.protobuf.Timestamp time = 3;
}

message DigitalInState {
	PinId pin = 1;
	bool state = 2;
	google.protobuf.Timestamp time = 3;
}

message NewByteData {
	PinId pin = 1;
	bytes data = 2;
	google.protobuf.Timestamp time = 3;
}

message DataWriteComplete {
	PinId pin = 1;
}

message InterfaceInitResult {
	Interfaces interface = 1;
	bool is_succeed = 2;
}

message InterfaceLost {
	Interfaces interface = 1;
}

message InterfaceReconnected {
	Interfaces interface = 1;
}

message DigitalInInitResult {
	Interfaces interface = 1;
	uint32 pin = 2;
	bool is_succeed = 3;
}
	
message DigitalOutInitResult {
	Interfaces interface = 1;
	uint32 pin = 2;
	bool is_succeed = 3;
}

message OnlyOneConnectionAllowed {
}

message LabNetResetReply {
	bool is_reset = 1;
}

message LabNetIdReply {
	string id = 1;
	uint32 major_version = 2;
	uint32 minor_version = 3;
}
// general messages


// digital out loop
message DigitalOutLoopStartResult {
	bool is_succeed = 1;
}

message DigitalOutLoopStopped {
	string loop_name = 1;
}
// digital out loop


message ServerWrappedMessage {
	oneof server_message {
		DigitalOutState digital_out_state = 1;
		DigitalInState digital_in_state = 2;
		NewByteData new_byte_data = 3;
		DataWriteComplete data_write_complete = 4;
		InterfaceInitResult interface_init_result = 5;
		DigitalInInitResult digital_in_init_result = 6;
		DigitalOutInitResult digital_out_init_result = 7;
		OnlyOneConnectionAllowed only_one_connection_allowed = 8;
		LabNetResetReply reset = 9;
		LabNetIdReply id = 10;
		InterfaceLost interface_lost = 11;
		InterfaceReconnected interface_reconnected = 12;
		DigitalOutLoopStartResult digital_out_loop_start_result = 13;
		DigitalOutLoopStopped digital_out_loop_stopped = 14;
	}
}